File: backend/models/notes.py
from pydantic import BaseModel, Field
from datetime import datetime
from typing import Optional, List, Dict, Any
from uuid import UUID

class FolderBase(BaseModel):
    name: str
    slug: str
    description: Optional[str] = None
    is_system: bool = False

class FolderCreate(FolderBase):
    pass

class FolderUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None

class FolderInDB(FolderBase):
    id: UUID
    user_id: Optional[UUID] = None
    created_at: datetime
    updated_at: datetime
    total_count: Optional[int] = None
    
    class Config:
        from_attributes = True

class NoteBase(BaseModel):
    folder_id: UUID
    title: str = "Untitled Note"
    content: Dict[str, Any] = Field(default_factory=dict)
    is_pinned: bool = False
    is_favorite: bool = False
    is_archived: bool = False
    metadata: Optional[Dict[str, Any]] = None

class NoteCreate(NoteBase):
    pass

class NoteUpdate(BaseModel):
    folder_id: Optional[UUID] = None
    title: Optional[str] = None
    content: Optional[Dict[str, Any]] = None
    is_pinned: Optional[bool] = None
    is_favorite: Optional[bool] = None
    is_archived: Optional[bool] = None
    metadata: Optional[Dict[str, Any]] = None

class NoteInDB(NoteBase):
    id: UUID
    user_id: Optional[UUID] = None
    is_deleted: bool = False
    deleted_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    version: int = 1
    content_preview: Optional[str] = None
    total_count: Optional[int] = None
    
    class Config:
        from_attributes = True

class NoteUpsertResponse(BaseModel):
    note_id: UUID
    was_created: bool

class DeleteResponse(BaseModel):
    success: bool
    message: str

class RestoreNoteRequest(BaseModel):
    """Request model for restoring a note"""
    note_id: UUID

# ==================== TAGS ====================

class TagBase(BaseModel):
    """Base tag model"""
    name: str
    color: Optional[str] = '#6B7280'

class TagCreate(TagBase):
    """Create tag model"""
    pass

class TagUpdate(BaseModel):
    """Update tag model"""
    name: Optional[str] = None
    color: Optional[str] = None

class TagInDB(TagBase):
    """Tag database representation"""
    id: UUID
    user_id: UUID
    created_at: datetime
    updated_at: datetime
    note_count: Optional[int] = 0

    class Config:
        from_attributes = True

class TagNoteRequest(BaseModel):
    """Request to tag/untag a note"""
    note_id: UUID
    tag_name: str
    tag_color: Optional[str] = None

class UntagNoteRequest(BaseModel):
    """Request to remove tag from note"""
    note_id: UUID
    tag_id: UUID

# ==================== TEMPLATES ====================

class TemplateBase(BaseModel):
    """Base template model"""
    name: str
    description: Optional[str] = None
    content: Dict[str, Any] = {"root": {"children": []}}

class TemplateCreate(TemplateBase):
    """Create template model"""
    pass

class TemplateUpdate(BaseModel):
    """Update template model"""
    name: Optional[str] = None
    description: Optional[str] = None
    content: Optional[Dict[str, Any]] = None

class TemplateInDB(TemplateBase):
    """Template database representation"""
    id: UUID
    user_id: Optional[UUID] = None
    is_system: bool = False
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

File: lib/types/notes.ts
/**
 * Frontend types for Notes API
 * Matches the backend models from backend/models/notes.py
 */

// ==================== FOLDERS ====================

export interface FolderBase {
  name: string;
  slug: string;
  description?: string | null;
  is_system: boolean;
}

export interface FolderCreate extends FolderBase {}

export interface FolderUpdate {
  name?: string;
  description?: string | null;
}

export interface Folder extends FolderBase {
  id: string;
  user_id?: string | null;
  created_at: string;
  updated_at: string;
  total_count?: number | null;
}

// ==================== NOTES ====================

export interface NoteBase {
  folder_id: string;
  title: string;
  content: Record<string, any>;
  is_pinned: boolean;
  is_favorite: boolean;
  is_archived: boolean;
  metadata?: Record<string, any> | null;
}

export interface NoteCreate extends NoteBase {}

export interface NoteUpdate {
  folder_id?: string;
  title?: string;
  content?: Record<string, any>;
  is_pinned?: boolean;
  is_favorite?: boolean;
  is_archived?: boolean;
  metadata?: Record<string, any> | null;
}

export interface Note extends NoteBase {
  id: string;
  user_id: string;
  is_deleted: boolean;
  deleted_at?: string | null;
  created_at: string;
  updated_at: string;
  version: number;
  content_preview?: string | null;
  total_count?: number | null;
}

export interface NoteUpsertResponse {
  note_id: string;
  was_created: boolean;
}

export interface DeleteResponse {
  success: boolean;
  message: string;
}

export interface RestoreNoteRequest {
  note_id: string;
}

// ==================== TAGS ====================

export interface TagBase {
  name: string;
  color?: string | null;
}

export interface TagCreate extends TagBase {}

export interface TagUpdate {
  name?: string;
  color?: string | null;
}

export interface Tag extends TagBase {
  id: string;
  user_id: string;
  created_at: string;
  updated_at: string;
  note_count?: number;
}

export interface TagNoteRequest {
  note_id: string;
  tag_name: string;
  tag_color?: string | null;
}

export interface UntagNoteRequest {
  note_id: string;
  tag_id: string;
}

// ==================== TEMPLATES ====================

export interface Template {
  id: string;
  user_id?: string;
  name: string;
  description?: string;
  content: any;
  is_system: boolean;
  created_at: string;
  updated_at: string;
}

export interface TemplateCreate {
  name: string;
  description?: string;
  content?: any;
}

export interface TemplateUpdate {
  name?: string;
  description?: string;
  content?: any;
}

// ==================== QUERY PARAMETERS ====================

export interface GetFoldersParams {
  search_term?: string;
  is_system?: boolean;
  limit?: number;
  offset?: number;
  sort_by?: 'name' | 'slug' | 'created_at' | 'updated_at';
  sort_order?: 'ASC' | 'DESC';
}

export interface GetNotesParams {
  note_id?: string;
  folder_slug?: string;
  search_term?: string;
  is_favorite?: boolean;
  is_pinned?: boolean;
  is_archived?: boolean;
  include_deleted?: boolean;
  limit?: number;
  offset?: number;
  sort_by?: 'title' | 'created_at' | 'updated_at' | 'is_pinned' | 'is_favorite';
  sort_order?: 'ASC' | 'DESC';
}

export interface SearchTagsParams {
  search_term: string;
  limit?: number;
}

// ==================== RESPONSE TYPES ====================

export interface ApiSuccessResponse {
  success: boolean;
  message?: string;
}

export interface CreateResponse {
  id: string;
  success: boolean;
}

export interface ToggleFavoriteResponse {
  is_favorite: boolean;
}
